mars/dataframe/base/astype.py

```diff
@@ -0,0 +1,309 @@
+                chunk_op = op.copy().reset_key()
+                params = c.params.copy()
+                dtypes = out.dtypes[cum_nsplits[c.index[1]]: cum_nsplits[c.index[1] + 1]]
+                params['dtypes'] = dtypes
+                chunk_category_cols = []
+                chunk_unique_chunks = []
+                for col in c.columns_value.to_pandas():
+                    if col in category_cols:
+                        chunk_category_cols.append(col)
+                        chunk_unique_chunks.append(unique_chunks[col])
+                chunk_op._category_cols = chunk_category_cols
+                new_chunk = chunk_op.new_chunk([c] + chunk_unique_chunks, **params)
+                out_chunks.append(new_chunk)
+        else:
+            for c in in_df.chunks:
+                chunk_op = op.copy().reset_key()
+                params = c.params.copy()
+                dtypes = out.dtypes[cum_nsplits[c.index[1]]: cum_nsplits[c.index[1] + 1]]
+                params['dtypes'] = dtypes
+                new_chunk = chunk_op.new_chunk([c], **params)
+                out_chunks.append(new_chunk)
+
+        new_op = op.copy()
+        return new_op.new_dataframes(op.inputs, nsplits=in_df.nsplits,
+                                     chunks=out_chunks, **out.params.copy())
+
+    @classmethod
+    def tile(cls, op):
+        if len(op.inputs[0].chunks) == 1:
+            return cls._tile_one_chunk(op)
+        elif isinstance(op.inputs[0], SERIES_TYPE):
+            return cls._tile_series(op)
+        else:
+            return cls._tile_dataframe(op)
+
+    @classmethod
+    def execute(cls, ctx, op):
+        in_data = ctx[op.inputs[0].key]
+        if not isinstance(op.dtype_values, dict):
+            if op.category_cols is not None:
+                uniques = [ctx[c.key] for c in op.inputs[1:]]
+                dtype = dict((col, CategoricalDtype(unique_values)) for
+                             col, unique_values in zip(op.category_cols, uniques))
+                ctx[op.outputs[0].key] = in_data.astype(dtype, errors=op.errors)
+
+            else:
+                ctx[op.outputs[0].key] = in_data.astype(op.dtype_values, errors=op.errors)
+        else:
+            selected_dtype = dict((k, v) for k, v in op.dtype_values.items()
+                                  if k in in_data.columns)
+            if op.category_cols is not None:
+                uniques = [ctx[c.key] for c in op.inputs[1:]]
+                for col, unique_values in zip(op.category_cols, uniques):
+                    selected_dtype[col] = CategoricalDtype(unique_values)
+            ctx[op.outputs[0].key] = in_data.astype(selected_dtype, errors=op.errors)
+
+    def __call__(self, df):
+        if isinstance(df, SERIES_TYPE):
+            self._object_type = ObjectType.series
+            empty_series = build_empty_series(df.dtype)
+            new_series = empty_series.astype(self.dtype_values)
+            dtype = CategoricalDtype() if isinstance(
```

Comment:
Convert to empty CategoricalDtype when the dtype is different from previous one. Consider the situation that the input df is created from pandas, and the dtype is just CategoricalDtype.